common/lib/xmodule/xmodule/js/src/video/09_completion.js

```diff
@@ -0,0 +1,144 @@
+                delete this.state.completionHandler;
+            },
+
+            /** Initializes the module. */
+            initialize: function() {
+                this.lastSentTime = undefined;
+                this.this.isComplete = false;
+                this.completionPercentage = this.state.config.completionPercentage;
+                this.startTime = this.state.config.startTime;
+                this.endTime = this.state.config.endTime;
+                this.isEnabled = this.state.config.completionEnabled;
+                if (this.endTime) {
+                    this.completeAfterSeconds = this.calculateCompleteAfterSeconds(this.startTime, this.endTime);
+                }
+                if (this.isEnabled) {
+                    this.bindHandlers();
+                }
+            },
+
+            /** Bind any necessary function callbacks to DOM events. */
+            bindHandlers: function() {
+                var self = this;
+
+                /** Event handler to check if the video is complete, and submit
+                 *  a completion if it is.
+                 *
+                 *  If the timeupdate handler doesn't fire after the required
+                 *  percentage, this will catch any fully complete videos.
+                 */
+                this.state.el.on('ended.completion', function() {
+                    self.handleEnded();
+                });
+
+                /** Event handler to check video progress, and mark complete if
+                 *  greater than completionPercentage
+                 **/
+                this.state.el.on('timeupdate.completion', function(ev, currentTime) {
+                    self.handleCompletion(currentTime);
+                });
+            },
+
+            handleEnded: function() {
+                if (this.isComplete) {
+                    return;
+                }
+                this.markCompletion();
+            },
+
+            handleCompletion: function(currentTime) {
+                var duration;
+                if (this.isComplete) {
+                    return;
+                }
+                if (this.lastSentTime !== undefined && currentTime - this.lastSentTime < this.repostDelaySeconds()) {
+                    // Throttle attempts to submit in case of network issues
+                    return;
+                }
+                if (this.completeAfterSeconds === undefined) {
+                    // Duration is not available at initialization time
+                    duration = this.state.videoPlayer.duration();
+                    if (!duration) {
+                        // duration is not yet set. Wait for another event,
+                        // or fall back to 'ended' handler.
+                        return;
+                    }
+                    this.completeAfterSeconds = this.calculateCompleteAfterSeconds(this.startTime, duration);
+                }
+
+                if (currentTime > this.completeAfterSeconds) {
```

Comment:
I had proposed `completeAfter` => `completeAfterTime`.  You instead went with `completeAfterSeconds`.

I find the code confusing. Since `currentTime` and `this.completeAfterSeconds`, they seem to contain the same unit.  I'd like to see then consistently named.  If they are both times, then please rename the latter to `this.completeAfterTime`.  If they are both in seconds from the beginning, then options might be:
- rename all the "xxxTime" variables to "xxxSeconds", or 
- drop a comment where they are defined, and just explain what they contain.

Or if I am missing something entirely, feel free to help clarify and make some other code change that makes this comparison more clear.  Thanks.

Reply:
@robrap I was thinking that time and seconds were different, analogous to the way datetime and timedelta are different, but it looks like you're right, and Javascript treats them identically.  I think you're right that they should all have the same suffix.  I'm leaning toward going with `xxxSeconds`, even though the names are slightly more awkward, because it makes the semantics clearer in all contexts (including the Javascript code, the Django settings modules, and JSON files that feed into the settings).

Reply:
@jcdyer What does seconds actually mean?  What are sample values for startTime (or startSeconds)?  Maybe you could clarify this with a quick comment if the name isn't enough to understand.

Reply:
On further look, I've settled on calling everything "Time."  I was hoping to distinguish between DateTime-like objects (which in javascript are represented as "seconds since the epoch", and the number of seconds from the beginning of a video, but even that distinction was going to end up muddied, because the video player exposes "startTime" and "endTime" (which are seconds since the start of the video), so as per your initial recommendation, I'm settling on a consistent use of "Time," with a comment to explain where meaning might be confusing.